spec.md

```diff
@@ -0,0 +1,1597 @@
+* `(relation) B @cascade`: deleting $`a : A`$ with existing $`b :_! B(a:I,...)`$, such that $`b :_! B(...)`$ violates $B$'s cardinality _for any role_ of $B$, triggers deletion of $b$.
+  * **defaults** to **TT** error
+* `(attribute) B @independent`. When deleting $`a : A`$ with existing $`b :_! B(a:O_B)`$, update the latter to $`b :_! B`$.
+  * **defaults** to: deleting $`a : A`$ with existing $`b :_! B(a:O_B)`$ triggers deletion of $b$.
+
+
+### Value types
+
+**Case PRIMITIVES**
+* `bool`
+  * Terms: `true`, `false`
+* `long` — _Comment: still think this could be named more nicely_
+  * Terms: 64bit integers
+* `datetime`
+  * TODO: list formats
+* `time` — _Comment: relative times_
+* `string` — _Comment: dynamically sized type; but could consider Arrow-like implementation, see e.g. [here](https://pola.rs/posts/polars-string-type/)_
+
+**Case STRUCT**
+```
+struct S:
+  C1 value V1 (@values(<EXPR>)),
+  C2 value V2 (@values(<EXPR>));
+```
+adds
+* _Struct type_ $S : \mathbf{Type}$
+* _Struct components_ $`C_1 : \mathbf{Type}`$, $`C_2 : \mathbf{Type}`$, and identify $`S = C_1 \times C_2`$
+* _Component value casting_: $`C_1 < V_1`$, $`C_2 < V_2`$
+* _Component value terms_: whenever $`v : V_i`$ then $`v : C_i`$ if $v$ conforms with `<EXPR>`
+  * **defaults** to: whenever $`v : V_i`$ then $`v : C_i`$ (no condition)
+* **Generalizes** to $n$ components
+
+### Functions defs
+
+**Case STREAM_RET_FUN**
+```
+fun F (x: T, y: S) -> { A, B }:
+  match <PATTERN>
+  (<OPERATORS>)
+  return { z, w };
+```
+adds the following to our type system:
+* _Function symbol_: $F : \mathbf{Type}(T,S)$.
+* _Function type_: when $`x : T`$ and $y: S$ then $F(x:T, y:S) : \mathbf{Type}$
+* _Output cast_: $F(x:T, y:S) < A \times B$
+* _Function terms_: $`(z,w) : F(x:T, y:S)`$ are discussed in section "Function semantics"
+* **Generalizes** to $n$ inputs and $m$ outputs
+
+**Case SINGLE_RET_FUN**
+```
+fun f (x: T, y: S) -> A, B:
+  match <PATTERN>
+  (<OPERATORS>)
+  return <AGG>, <AGG>;
+```
+adds the following to our type system:
+* _Function symbol_: when $`x : T`$ and $y: S$ then $f(x:T, y:S) : A \times B$
+* _Function terms_: $`(z,w) : f(x:T, y:S)`$ are discussed in section "Function semantics"
+* **Generalizes** to $n$ inputs and $m$ outputs
+
+_Comment: notice difference in capitalization between the two cases!_
+
+## Undefine semantics
+
+### Type axioms
+
+**Case ENT**
+* `entity A` removes $`A : \mathbf{Ent}`$
+* `sub B from (entity) A` removes $`A < B`$
+
+**Case REL**
+* `relation A` removes $`A : \mathbf{Rel}`$
+* `sub B from (relation) A` removes $`A < B`$
+* `relates I from (relation) A` removes $`A : \mathbf{Rel}(I)`$
+* `as J from (relation) A relates I` removes $`I <_! J`$ 
+* `relates I[] from (relation) A` removes $`A : \mathbf{Rel}([I])$
+* `as J[] from (relation) A relates I[]` removes $`I <_! J`$
+
+**Case ATT**
+* `attribute A` removes $`A : \mathbf{Att}`$ and $`A : \mathbf{Att}(O_A)`$
+* `value V from (attribute) A value V` removes $`A < V`$
```

Comment:
```suggestion
* `value V from (attribute) A` removes $`A < V`$
```

Reply:
Additionally, this raises the larger question of how TypeDB should deal with a valueless attribute type within a transaction. Should undefine be illegal for this (only redefine permitted)? Should this be a commit-time error? The spec should be specific as to these details, in this case and in any others similar.